enhancements/workload-partitioning/wide-availability-workload-partitioning.md

```diff
@@ -0,0 +1,731 @@
+apiVersion: performance.openshift.io/v2
+kind: PerformanceProfile
+metadata:
+  name: openshift-node-workload-partitioning-custom
+spec:
+  cpu:
+    isolated: 2-3
+    reserved: 0,1
+    # New enum addition
+    workloads:
+      - Infrastructure
+```
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Changes to NTO
+
+The NTO PerformanceProfile will be updated to support a new flag which will
+toggle the workload pinning to the `reserved` cores. The idea here being to
+simplify the approach for how customers set this configuration. With PAO being
+part of NTO now ([see here for more info](../node-tuning/pao-in-nto.md)) this
+affords us the chance to consolidate the configuration for `kubelet` and `crio`.
+
+We will modify the code path that generates the [new machine
+config](https://github.com/openshift/cluster-node-tuning-operator/blob/a780dfe07962ad07e4d50c852047ef8cf7b287da/pkg/performanceprofile/controller/performanceprofile/components/machineconfig/machineconfig.go#L91-L127)
+using the performance profile. With the new `spec.workloads[Infrastructure]`
+enum we will add the configuration for `crio` and `kubelet` to the final machine
+config manifest. Then the existing code path will apply the change as normal.
+
+#### API Server Admission Hook
+
+We will need to alter the code in the admission controller to remove the check
+for Single Node Topology, and modify the check for running nodes to check the
+global identifier which will be set at install time.
+
+The existing admission hook has 4 checks when it comes to workload pinning.
+
+Old Path:
+
+1. Check if `pod` is a static pod
+   - Skips modification attempt if it is static.
+2. Checks if currently running cluster topology is Single Node - **Will Change**
+   - Skips modification if it is anything other than Single Node
+3. Checks if all running nodes are managed - **Will Change**
+   - Skips modification if any of the nodes are not managed
+4. Checks what resource limits and requests are set on the pod
+   - Skips modification if QoS is guaranteed or both limits and requests are set
+   - Skips modification if after update the QoS is changed
+
+Changed Path:
+
+1. Check if `pod` is a static pod
+   - Skips modification attempt if it is static.
+2. Checks if currently running cluster has global identifier for partitioning
+   set
+   - Skips modification if identifier partitioning set to `None` unless Single
+     Node, will check with old logic to maintain upgrade for Single-Node 4.11 ->
+     4.12.
+3. Checks what resource limits and requests are set on the pod
+   - Skips modification if QoS is guaranteed or both limits and requests are set
+   - Skips modification if after update the QoS is changed
+
+### Risks and Mitigations
+
+The sames risks and mitigations highlighted in [Management Workload
+Partitioning](management-workload-partitioning.md) apply to this enhancement as
+well.
+
+We need to make it very clear to customers that this feature is supported as a
```

Comment:
PerformanceProfiles are not cluster wide, but scoped to MCPs. You can (and will) have multiple PerfProfiles in a single cluster. Will all of them have to use the same configuration of workload partitioning? What about someone using this on non-SNO cluster?

Reply:
The idea for this is driven by that MCP link you mentioned, so the way this currently works is that you can have master/worker/custom_pool PerformanceProfiles and those configurations will be applied to those specific nodes, so you can have a heterogenous cluster, as long as the MCPs are homogenous there should be no concerns about about multiple PerfProfiles. In my testing I had a 2CPUSet part for masters and a 1CPUSet part for workers for example.

Reply:
CPU sets can differ indeed. But what about the proposed workloads api values?

Reply:
If I understand your question right Martin, the values in this current design iteration are retrieved in the reserved/isolated values, so the PerfProfile with `workloads[Management]` set we will use those values in `reserved` to create the specific MCP config for `crio/kublet`, if no workloads is set, then `crio/kubelet` won't be updated. (I'm sure we'll discuss a different approach in our meeting)

Reply:
After our discussion, it was decided that for this enhancement we would be changing the PerformanceProfile API